src/System.Private.CoreLib/src/System/Threading/ThreadPool.cs

```diff
@@ -1342,22 +1341,41 @@ public static bool UnsafeQueueUserWorkItem(WaitCallback callBack, object state)
 
             EnsureVMInitialized();
 
-            IThreadPoolWorkItem tpcallBack = new QueueUserWorkItemCallback(callBack, state, null);
+            object tpcallBack = new QueueUserWorkItemCallback(callBack, state, null);
 
             ThreadPoolGlobals.workQueue.Enqueue(tpcallBack, forceGlobal: true);
 
             return true;
         }
 
-        internal static void UnsafeQueueCustomWorkItem(IThreadPoolWorkItem workItem, bool forceGlobal)
+        public static bool UnsafeQueueUserWorkItem(IThreadPoolWorkItem callBack, bool preferLocal)
         {
-            Debug.Assert(null != workItem);
+            if (callBack == null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.callBack);
+            }
+            if (callBack is Task)
```

Comment:
Since `Task` is no longer an `IThreadPoolWorkItem`, is this check still necessary?

Reply:
> Since Task is no longer an IThreadPoolWorkItem, is this check still necessary?

Unfortunately, a developer could still do:
```C#
public class MyTask : Task, IThreadPoolWorkItem
{
    public MyTask(Action action) : base(action) { }
}
```
and if they then did:
```C#
var t = new MyTask();
t.Start();
ThreadPool.UnsafeQueueUserWorkItem(t, true);
```
it could end up running through the code in the task twice, which will break invariants in the task code.

We could decide we don't care about this corner-case (it's super rare to derive from Task like this), but I put the check in just to be safe.

Reply:
Ah ok